test/cpp/naming/resolver_component_test.cc

```diff
@@ -0,0 +1,338 @@
+              "'<ip0:port0>,<is_balancer0>;<ip1:port1>,<is_balancer1>;...' "
+              "addresses of "
+              "backend and/or balancers. 'is_balancer' should be bool, i.e. "
+              "true or false.");
+DEFINE_string(expected_chosen_service_config, "",
+              "Expected service config json string that gets chosen (no "
+              "whitespace). Empty for none.");
+DEFINE_string(
+    local_dns_server_address, "",
+    "Optional. This address is placed as the uri authority if present.");
+DEFINE_string(expected_lb_policy, "",
+              "Expected lb policy name that appears in resolver result channel "
+              "arg. Empty for none.");
+
+namespace {
+
+class GrpcLBAddress final {
+ public:
+  GrpcLBAddress(std::string address, bool is_balancer)
+      : is_balancer(is_balancer), address(address) {}
+
+  bool operator==(const GrpcLBAddress &other) const {
+    return this->is_balancer == other.is_balancer &&
+           this->address == other.address;
+  }
+
+  bool operator!=(const GrpcLBAddress &other) const {
+    return !(*this == other);
+  }
+
+  bool is_balancer;
+  std::string address;
+};
+
+bool ConvertStringToBool(std::string bool_str) {
+  if (gpr_stricmp(bool_str.c_str(), "true") == 0) {
+    return true;
+  } else if (gpr_stricmp(bool_str.c_str(), "false") == 0) {
+    return false;
+  } else {
+    gpr_log(GPR_ERROR,
+            "invalid expected_addrs_list entry: expected false or true, got %s",
+            bool_str.c_str());
+    abort();
+  }
+}
+
+vector<GrpcLBAddress> ParseExpectedAddrs(std::string expected_addrs) {
+  std::vector<GrpcLBAddress> out;
+
+  while (expected_addrs.size() != 0) {
+    // get the next <ip>,<port> (v4 or v6)
+    size_t next_comma = expected_addrs.find(",");
+    if (next_comma == std::string::npos) {
+      gpr_log(
+          GPR_ERROR,
+          "Missing ','. Expected_addrs arg should be a semi-colon-separated "
+          "list of "
+          "<ip-port>,<bool> pairs. Left-to-be-parsed arg is |%s|",
+          expected_addrs.c_str());
+      abort();
+    }
+    std::string next_addr = expected_addrs.substr(0, next_comma);
+    expected_addrs = expected_addrs.substr(next_comma + 1, std::string::npos);
+    // get the next is_balancer 'bool' associated with this address
+    size_t next_semicolon = expected_addrs.find(";");
+    bool is_balancer =
+        ConvertStringToBool(expected_addrs.substr(0, next_semicolon));
+    out.emplace_back(GrpcLBAddress(next_addr, is_balancer));
+    if (next_semicolon == std::string::npos) {
+      break;
+    }
+    expected_addrs =
+        expected_addrs.substr(next_semicolon + 1, std::string::npos);
+  }
+  if (out.size() == 0) {
+    gpr_log(GPR_ERROR,
+            "expected_addrs arg should be a comma-separated list of "
+            "<ip-port>,<bool> pairs");
+    abort();
+  }
+  return out;
+}
+
+gpr_timespec TestDeadline(void) {
+  return grpc_timeout_seconds_to_deadline(100);
+}
+
+typedef struct ArgsStruct {
```

Comment:
Nit: Since this is C++, the `typedef` is not needed.  Once you declare the struct, you can refer to it by name directly.

Reply:
done